File: app/nilink/lib/util/DataProvider.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:http/http.dart';
import 'package:nilink/main.dart';
import 'package:nilink/util/Auth.dart';

class DataProvider {
  static String url = "https:/nilink.cf";

  static Map<String,String> endpoints = <String,String>{
    "applications":"/api/user/applications",
    "application":"/api/user/application/{id}",
  };

  static Future<Map<String, dynamic>> request(
    String endpoint, dynamic params, String token) async {
    Response r = await http.post(Uri.parse(url + endpoint), body: params, headers: {
      'Accept': 'application/json',
      'Authorization': 'Bearer $token',
    });
    bool error = r.statusCode != 200;
    dynamic result = jsonDecode(r.body);
    int statusCode = r.statusCode;
    if (r.statusCode == 401 && token!=null) {
      dynamic result = await Auth.refreshToken();
      if (!result["error"]) {
        return await request(endpoint, params, Auth.storage.getString("token"));
      } else {
        Auth.deleteToken();
        Navigator.pushNamedAndRemoveUntil(navigatorKey.currentContext, '/login', (route) => false);
      }
    }
    return <String, dynamic>{
      "error": error,
      "result": result,
      "status": statusCode
    };
  }


static Future<Map<String, dynamic>> getRequest(
    String endpoint, String token) async {
    Response r = await http.get(Uri.parse(url + endpoint), headers: {
      'Accept': 'application/json',
      'Authorization': 'Bearer $token',
    });
    bool error = r.statusCode != 200;
    dynamic result = jsonDecode(r.body);
    int statusCode = r.statusCode;
    if (r.statusCode == 401) {
      dynamic result = await Auth.refreshToken();
      if (!result["error"]) {
        return await getRequest(endpoint, Auth.storage.getString("token"));
      } else {
        Auth.deleteToken();
        Navigator.pushNamedAndRemoveUntil(navigatorKey.currentContext, '/login', (route) => false);
      }
    }
    return <String, dynamic>{
      "error": error,
      "result": result,
      "status": statusCode
    };
  }

  static Future<Map<String,dynamic>> getApplications() async {
    return await DataProvider.getRequest(endpoints["applications"], Auth.token);
  }

  static Future<Map<String,dynamic>> getApplication(String id) async {
    return await DataProvider.getRequest(endpoints["application"].replaceAll("{id}", id), Auth.token);
  }
}

File: web/dev/js/DataProvider.js
class DataProvider{

    static url="https://nilink.cf";

    static endopoints={
        'applications': "/api/developer/applications",
        'application': "/api/developer/application/{id}",
        'create-application': "/api/developer/application/create",
        'update-application': "/api/developer/application/{id}/update",
        'delete-application' : "/api/developer/application/{id}/delete",
        'upload-icon' : "/api/developer/application/{id}/upload/icon",
        'upload-screen' : "/api/developer/application/{id}/upload/screenshot",
        'delete-screen' : "/api/developer/application/{id}/screenshot/{screen}/delete"
    }


    static async post(url, token='', data = {}, method='POST') {
        const response = await fetch(url, {
          method: method,
          headers: {
            'Authorization': 'Bearer ' + token
          },
          redirect: 'follow', 
          body: data
        });
        var res=await response.json(); 
        if(response.status==401){
          var result = await Auth.tryLogin();
          if(result){
            return await this.post(url,Auth.token,data);
          }else{
            Auth.logout();
          }
        }
        return {
            'data':res,
            'code':response.status,
            'error':response.status!=200
        }
    }

    static async get(url, token='') {
        const response = await fetch(url, {
          method: 'GET',
          headers: {
            'Authorization': 'Bearer ' + token
          },
          redirect: 'follow', 
          referrerPolicy: 'no-referrer', 
        });
        var res=await response.json(); 
        console.log(res);
        if(response.status==401){
          var result = await Auth.tryLogin();
          if(result){
            return await this.get(url,Auth.token);
          }else{
            Auth.logout();
          }
        }
        return {
            'data':res,
            'code':response.status,
            'error':response.status!=200
        }
    }

    static async getApplications(callback){
      this.get(this.endopoints["applications"], Auth.token).then(callback);
    }

    static async getApplication(idApp, callback){
      console.log(idApp);
      this.get(this.endopoints["application"].replace("{id}",idApp), Auth.token).then(callback);
    }

    static async createApplication(data,callback){
      this.post(this.endopoints["create-application"], Auth.token, data).then(callback);
    }

    static async updateApplication(idApp,data,callback){
      this.post(this.endopoints["update-application"].replace("{id}",idApp), Auth.token, data, 'PATCH').then(callback);
    }

    static async deleteApplication(idApp,callback){
      this.post(this.endopoints["delete-application"].replace("{id}",idApp), Auth.token, {}, 'DELETE').then(callback);
    }

    static async uploadIcon(idApp,data,callback){
      this.post(this.endopoints["upload-icon"].replace("{id}",idApp), Auth.token, data, 'POST').then(callback);
    }

    static async uploadScreen(idApp,data,callback){
      this.post(this.endopoints["upload-screen"].replace("{id}",idApp), Auth.token, data, 'POST').then(callback);
    }

    static async deleteScreen(idApp, idScreen,callback){
      this.post(this.endopoints["delete-screen"].replace("{id}",idApp).replace("{screen}",idScreen), Auth.token, {}, 'DELETE').then(callback);
    }
    

}